BranchWatchDialog: A Total Replacement for CodeDiffDialog

mitaclaw opened the pull request:
With a purpose-built Branch Watch feature built into the emulated system, BranchWatchDialog (formerly named CodeDiffDialog) is now better than ever!

sepalani commented:
@dolphin-emu-bot rebuild

mitaclaw commented:
Darn it, parenthesiszed initialization of aggregates has bad compiler support.
https://en.cppreference.com/w/cpp/20

AdmiralCurtiss commented:
This is very cool but also jesus why is this is a single 2000 loc commit

Please at the very least break out the Core and JIT changes into separate things. The JIT stuff in particular seems like it could have a performance impact even when the code diff is not in use so that needs scrutiny.

mitaclaw commented:
I can split this up if it's too much to review all at once.

All JIT'd code for Branch Watch will only emit when m_debugging_enabled is true, so for the non-debugging user the only performance draw is a single condition when recompiling branch instructions. Even still, the emitted code also checks if Branch Watch is enabled before doing the expensive ABI_CallFunction operation and std::unordered_map lookup. There is a 50-66% performance reduction when Branch Watch is enabled in the JIT, but that's sort of expected from this kind of debugging feature.  I do have some ideas for optimization, but for now this simple implementation works great.

mitaclaw commented:
There's two ways I can see this being split:
PR 1) BranchWatch + CPUCore changes (+612 -42)
PR 2) DolphinQt changes (+1328 -611)
or:
PR 1) BranchWatch addition (+448 -0)
PR 2) CPUCore changes (+164 -42)
PR 3) DolphinQt changes (+1328 -611)

Which way would you think is better? Either way, there is a massive loc change with the DolphinQt split because I wrote three new Qt classes (CodeDiffTableModel, CodeDiffProxyModel, CodeDiffTableView) and basically completely rewrote another (CodeDiffDialog) and they're all intertwined.

AdmiralCurtiss commented on Source/Core/Core/PowerPC/Jit64/Jit_Branch.cpp:
```diff
@@ -94,6 +116,8 @@ void Jit64::bx(UGeckoInstruction inst)
   gpr.Flush();
   fpr.Flush();
 
+  WriteBranchWatch(js.compilerPC, Imm32(js.op->branchTo), inst);
```
As far as I can tell, this is always emitted, even outside of debugging mode.

AdmiralCurtiss commented:
To be clear it's fine if this is a single PR, but it should be multiple commits. And I'd prefer more commits rather than less.

mitaclaw commented:
So as it turns out, I had genuinely forgotten what the performance overhead of the old Code Diff Tool was like.  For me (testing Animal Crossing and Mario Kart: Double Dash), the new Code Diff Tool using Branch Watch is no less performant, maybe even slightly faster, than the old one using the JIT profiling data.

lioncash commented on Source/Core/DolphinQt/Debugger/CodeDiffProxyModel.h:
```diff
@@ -0,0 +1,67 @@
+
+  QString m_symbol_name;
+  std::optional<u32> m_origin_min, m_origin_max, m_destin_min, m_destin_max;
+  bool m_b, m_bl, m_bc, m_bcl, m_blr, m_blrl, m_bclr, m_bclrl, m_bctr, m_bctrl, m_bcctr, m_bcctrl;
```
These should have initial default values for deterministic initialization. Also maybe put some of these on new lines, this is quite a bit in sequence (ditto for the optionals).

mitaclaw replied:
- The std::optional members will always be std::nullopt when default constructed.
- The bool members are initialized in `CodeDiffDialog::CreateLayout()` when the QCheckBoxes are connected and checked (in that order).
- In retrospect, I think the default construction of the QString member explains why I needed to check `QString::isNull()` in `CodeDiffProxyModel::filterAcceptsRow`.  I will go ahead and give the QString a default value.

lioncash replied:
I'm saying they should be defaulted so on construction they have a valid state. Which is good because it prevents weird uninitialized/undefined behavior from happening if refactors happen, and for whatever reason, any of them are accidentally touched before the call is made (this is a reasonable request).

I know the optional members default to a default valid state. I was more referring to line breaking, rather than initializing those.

lioncash commented on Source/Core/DolphinQt/Debugger/CodeDiffProxyModel.h:
```diff
@@ -0,0 +1,67 @@
+    this->*member = enabled;
+    invalidateRowsFilter();
+  }
+  void OnSymbolTextChanged(const QString& text)
```
```suggestion
  void OnSymbolTextChanged(QString text)
```

and std::move into `m_symbol_name`

mitaclaw replied:
This method is a slot for the signal, `QLineEdit::textChanged(const QString& text)`, and isn't needed for anything else. I would rather keep the signature matching, as it makes a lambda in `CodeDiffDialog::CreateLayout()` simple to reuse.

lioncash commented on Source/Core/DolphinQt/Debugger/CodeDiffTableModel.cpp:
```diff
@@ -0,0 +1,268 @@
+    return QVariant();
+
+  static const std::array<QString, Column::NumberOfColumns> headers = {
+      tr("Instr."),      tr("Origin"),     tr("Destination"),
```
```suggestion
      tr("Instruction"),      tr("Origin"),     tr("Destination"),
```

Since none of the others are abbreviated

mitaclaw replied:
The instruction column never needs to be as wide as the full word "instruction" is.  The CodeViewWidget is the same way.

lioncash commented on Source/Core/DolphinQt/Debugger/CodeDiffTableView.cpp:
```diff
@@ -0,0 +1,140 @@
+    [[fallthrough]];
+  case Column::Origin:
+  case Column::Destination:
+    m_code_widget->SetAddress(v.value<u32>(), CodeViewWidget::SetAddressUpdate::WithDetailedUpdate);
```
please use break, even if they're the last case in a switch

lioncash commented on Source/Core/DolphinQt/Debugger/CodeWidget.h:
```diff
@@ -17,6 +16,7 @@ class QSplitter;
 class QListWidget;
 class QPushButton;
 class QTableWidget;
+class CodeDiffDialog;
```
C goes above Q

mitaclaw commented on Source/Core/Core/PowerPC/BranchWatch.h:
```diff
@@ -0,0 +1,153 @@
+
+namespace Core
+{
+inline bool operator==(const BranchWatchKey& lhs, const BranchWatchKey& rhs) noexcept
```
As of C++20, the inequality operator is automatically generated by the compiler if operator== is defined.

mitaclaw commented on Source/Core/Core/PowerPC/BranchWatch.h:
```diff
@@ -0,0 +1,153 @@
+class BranchWatch
+{
+  // The JIT(s) access m_recording_active directly to test if it is true.
+  friend Jit64;
```
The JIT must get offsetof(Core::BranchWatch, m_recording_active) to access it from recompiled code without the ABI_CallFunction overhead.  However, what I can do is write a friend method to return this value so the JIT doesn't need to be a friend of BranchWatch.

lioncash replied:
Yes, either way would be preferable to friending the JITs. Though I sincerely doubt the overhead would be that much.

mitaclaw commented on Source/Core/Core/PowerPC/BranchWatch.cpp:
```diff
@@ -0,0 +1,291 @@
+// Copyright 2023 Dolphin Emulator Project
```
I would like a second opinion on this before moving it.

TryTwo replied:
Is BranchWatch passively collecting info for the debug widget?  At a glance most of it looked that way, so I would lean to Core/Debugger.  What were your thoughts on the location?

lioncash replied:
BreakPoints.h/.cpp should also be in the debug directory, but just haven't been moved there yet.

mitaclaw commented:
A thought just crossed my mind. This tool (monitoring branch instruction hits) doesn't really do the exact same thing as the old tool (monitoring JIT blocks associated with the starts of symbols). Given that, would it be better for it and the source code files to be named something like "BranchDiff"? The git diff for the new CodeDiffDialog.cpp/h is pretty much nonsense, as everything has been rewritten and shuffled around.

TryTwo commented:
I like the name branch watch to describe it.

mitaclaw commented:
So do you think the QDialog should be renamed BranchWatchDialog as well?

TryTwo commented:
Yeah, I'm okay with it changing. Using Branch in the name is more descriptive.

lioncash commented on Source/Core/Core/Debugger/BranchWatch.cpp:
```diff
@@ -0,0 +1,302 @@
+{
+  if (!CanSave())
+  {
+    ASSERT_MSG(CORE, false, "BranchWatch can not be saved.");
```
Looking over this again, Is invoking the assertion handler (which can kill dolphin if the panic handler is disabled) necessary over either returning an error code and letting the frontend handle it (usually with a localized and readable message), or printing an error log message.

mitaclaw replied:
The BranchWatchDialog already checks `bool Core::BranchWatch::CanSave()` before attempting to save, this was just an extra safety check to future-proof the component against misuse.  Do you think an ASSERT is overkill for this?  I can switch to a log message and return false.

lioncash replied:
In that case (I missed the check in the dialog implementation), I think this is perfectly fine

lioncash commented on Source/Core/Core/PowerPC/CachedInterpreter/CachedInterpreter.h:
```diff
@@ -50,6 +50,7 @@ class CachedInterpreter : public JitBase
                                                  UGeckoInstruction data);
   static void WritePC(CachedInterpreter& cached_interpreter, UGeckoInstruction data);
   static void WriteBrokenBlockNPC(CachedInterpreter& cached_interpreter, UGeckoInstruction data);
+  friend void DoBranchWatchHit(CachedInterpreter& cached_interpreter, UGeckoInstruction n);
```
Is this intended to be static void, rather than friend void?

mitaclaw replied:
`CachedInterpreter::DoBranchWatchHit` accesses the protected member, `JitBase::m_code_buffer`.  If this were an inline definition, I could just use static.  However, non-member functions defined outside of the class must use the friend declaration to access private members.
https://en.cppreference.com/w/cpp/language/friend

lioncash replied:
Right, but, why does it need to be a friend function when a static member would do the [same](https://godbolt.org/z/jxrMxj9qf) thing and also be consistent with the rest of the identifiers its around?

Like, unless I'm overlooking something obvious at 3AM, they should be essentially equivalent.

mitaclaw replied:
Man, I have no clue what the difference between your Compiler Explorer example and the code in CachedInterpreter is, but I get `Error	C2248	'JitBase::m_code_buffer': cannot access protected member declared in class 'JitBase' ` from changing it from friend to static. I thought the info on cppreference sounded cut-and-dry like what I needed.

mitaclaw replied:
Wait, yes I do.

mitaclaw commented:
I guess while the subject of moving is on the table, is BranchWatchDialog happy where it currently is? Would it better fit under the Tools menu on the menubar, next to the Cheats Manager? Or maybe it would be better as a floating widget like the MemoryWidget / CodeWidget? It functions fine where it is now, but accessing it is unlike any other debug tool.

mitaclaw commented:
If it remains a dialog window, one compromise could be having it accessible from both the CodeWidget and the MenuBar.

TryTwo commented:
Yeah the button is in a weird place.  I don't think it needs to become its own full widget, but others might have a different opinion.  

Placing it alongside cheat manager seems logical, but it'd have to be disabled when debugging is not enabled,. I'm not sure how people would feel about that.

mitaclaw commented:
The BranchWatchDialog may be undergoing a change to round off a rough corner.  Stay tuned.

mitaclaw commented:
At @TryTwo's request, and upon further reflection, I have revised the controls of the Branch Watch Tool to be a bit more like those from the original Code Diff Tool.
Previously, the Branch Watch Tool had Start / Stop and Pause / Resume buttons. The "Stop Recording" button would clear Branch Watch automatically, which TryTwo thought was unintuitive. Now, the tool has Start / Pause buttons and a Clear Branch Watch button.  Start / Pause is basically the same as the old Pause / Resume button. Branch Watch has been simplified to no longer have a "Disabled" phase. This removal has eliminated invalid states when Branch Watch is disabled, and also removed paranoid checks to make sure it is not disabled. A new feature has been added to wipe recent hits. Wiping recent hits may be useful if you are unsure if a branch has been hit or not, or to wipe away recent hits accidentally included in a Branch Watch snapshot file. This button replaces the "Restart Without Clear" button, which was basically worthless.
Finally, I have added a new feature of my own to the Tool menu. It is now possible to switch between showing the symbol at the origin address or destination address of a hit.

mitaclaw commented:
![Branch Watch Tool 1](https://github.com/dolphin-emu/dolphin/assets/140017135/df2cdada-926b-4223-94ac-7368ed52876e)

TryTwo commented:
@dolphin-emu-bot rebuild

TryTwo commented:
Thanks for working on my suggestions. Everything is feeling good for me.

dreamsyntax commented:
I finally was able to give this a try.
First impression: wow. This is a huge improvement.


I still need to use it a some morein real world use, but these are small items I noticed:

- The Help popups no longer have newlines, this makes it a bit hard to read, adding some newlines would help between relevant sections.
- The Help text incorrectly references "Start Recording" rather than current UI labels - overall the text should be re-checked/written to current.
- I really would like an indicator that I have previously blr'd a function, especially for save/load purposes.
- In my lists I sometimes saw invalid addresses (0x00000576 as an example)


That said, if the Help text was updated to reflect current UI I would be fine merging this as is, and adding BLR/Inspected item support later.

mitaclaw commented:
My bad on missing the outdated text, I will fix that and the newlines right away. The addresses such as 0x00000576 are not invalid, those are branches in exception handlers. Exception handlers execute in the physical address space with address translation disabled. I omitted the "Inspected" column from the Code Diff Tool in the creation of the new Branch Watch Tool because I thought it would be difficult to keep it up-to-date with current data. especially now with the multitude of ways a candidate can be "inspected" (NOP at origin, BLR at destination, BLR at symbol start). However, I can try and take a stab at it if it would mean leaving nothing behind in the transition.

mitaclaw commented:
![Branch Watch Inspected](https://github.com/dolphin-emu/dolphin/assets/140017135/7da3f041-c270-42d8-8cc8-44434d67ce40)
At @dreamsyntax's request, I have reintroduced and revamped the "inspected" details from the former Code Diff Tool.

Inserting a NOP at an origin address will mark that origin address cell as inspected. Inserting a BLR is more complex.

When inserting a BLR at the destination address, all cells with a matching destination address will be marked as inspected. When symbols are available, all cells with a symbol with a starting address matching the destination address will be marked as inspected.

When inserting a BLR at a symbol start, the behavior is basically the inverse. All cells with matching symbols will be marked as inspected.  Additionally, all of the destination address cells matching the starting address of the symbol will be marked as inspected.

Inspection is marked by a cell's text turning bold and red.

mitaclaw commented:
Parenthesized initialization of aggregates strikes again.

dreamsyntax commented:
@dolphin-emu-bot rebuild

mitaclaw commented:
Rebasing on latest to hopefully fix the mtl-osx fifo diffs.

dreamsyntax commented:
@dolphin-emu-bot rebuild

dreamsyntax commented:
@dolphin-emu-bot rebuild

dreamsyntax commented:
@dolphin-emu-bot rebuild

mitaclaw commented:
Changed the Wipe Recent Hits button to only be enabled when in the Reduction phase, as it had no function in the Blacklist phase.

dreamsyntax commented:
@dolphin-emu-bot rebuild

dreamsyntax reviewed (approved):
Just tried the latest variant, it meets all my expectations.

As for the code, what are the plans with the TODOs you've left in, given the functionality is working as expected?
Additionally, this won't work on ARM builds if using Jit?

mitaclaw commented:
The TODOs are things I was dissatisfied with, but are outside of the scope of the current PR (or are compiler vendor limitations). Some time in the future, maybe I or someone else will pick them up and make them better.
I had skipped on implementing BranchWatch on JitArm64 because the Branch Watch Tool is unavailable on Android, but I forgot Windows on ARM is also a platform Dolphin supports.  I might need some help verifying an ARM implementation works as expected.

JosJuice commented:
> I had skipped on implementing BranchWatch on JitArm64 because the Branch Watch Tool is unavailable on Android, but I forgot Windows on ARM is also a platform Dolphin supports.

Dolphin supports AArch64 on all OSes it supports. Currently the most popular OS for AArch64 aside from Android is macOS.

mitaclaw commented on Source/Core/Core/PowerPC/Jit64/Jit_Integer.cpp:
```diff
@@ -394,30 +394,42 @@ void Jit64::DoMergedBranch()
-      destination = SignExt16(next.BD << 2);
-    else
-      destination = nextPC + SignExt16(next.BD << 2);
+    const u32 destination = SignExt16(next.BD << 2) + !next.AA * nextPC;
```
Oops, I fixed every place I had changed this except for the one you actually referenced.

mitaclaw commented:
Bugs fixed (using ABI_PARAM1 and RSCRATCH2 improperly in Jit64), code optimized (less RegCache flushes, parameter packing added, BranchWatch members rearranged for JIT code emission), and JitArm64 support added.
I've come to realize that I actually barely know what I'm doing when writing code in either JIT, so I would really appreciate someone more experienced weighing in on if what I've written is at all correct and/or maximally efficient.

JosJuice commented on Source/Core/Core/PowerPC/Jit64/Jit.h:
```diff
@@ -98,6 +98,12 @@ class Jit64 : public JitBase, public QuantizedMemoryRoutines
   void WriteIdleExit(u32 destination);
+  void WriteBranchWatch(u32 origin, u32 destination, UGeckoInstruction inst,
+                        BitSet32 registers_in_use);
+  void WriteBranchWatchDestInRSCRATCH(u32 origin, UGeckoInstruction inst);
```
While the fact that this function doesn't have a `registers_in_use` parameter does provide a hint, I think it's not obvious that this function doesn't do any saving of registers. I assume you left out the `register_in_use` parameter because no called needed to save any registers other than RSCRATCH? Personally I think the intent of the code would become clearer at the call site if the function has a `registers_in_use` parameter that the caller populates with just RSCRATCH.

mitaclaw replied:
I agree. Thanks.

JosJuice commented on Source/Core/Core/PowerPC/Jit64/Jit_LoadStore.cpp:
```diff
@@ -228,6 +230,37 @@ void Jit64::lXXx(UGeckoInstruction inst)
+static_assert((BitSet32{ABI_PARAM1} & ABI_ALL_CALLER_SAVED) != BitSet32{});
+#endif
+
+// RSCRATCH2 will hold the amount of faked branch watch hits (loop count minus one)
```
Is this a precondition or a postcondition? It reads like it's a postcondition, but as far as I can tell from reading the code it's actually a precondition.

Also, like with WriteBranchWatchDestInRSCRATCH, if the caller would add RSCRATCH2 to the bitset instead of the function doing it internally, I think the code would be clearer.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/Jit.h:
```diff
@@ -315,6 +315,14 @@ class JitArm64 : public JitBase, public Arm64Gen::ARM64CodeBlock, public CommonA
+  void WriteBranchWatch(u32 origin, u32 destination, UGeckoInstruction inst, BitSet32 gprs_in_use,
+                        BitSet32 fprs_in_use);
+  void WriteBranchWatchDestInScratch(u32 origin, Arm64Gen::ARM64Reg destination,
+                                     UGeckoInstruction inst);
```
Same comments here as for Jit64.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_Branch.cpp:
```diff
@@ -74,6 +76,51 @@ void JitArm64::rfi(UGeckoInstruction inst)
+  const ARM64Reg WA = gpr.GetReg();
+  LDRB(IndexType::Unsigned, WA, branch_watch,
+       static_cast<int>(Core::BranchWatch::GetOffsetOfRecordingActive()));
+  FixupBranch branch = TBNZ(WA, true);
```
Using `CBNZ` instead of `TBNZ` would be a bit more idiomatic. I don't think it actually affects performance, though.

mitaclaw replied:
CBZ, actually. I'm not sure how it was working with TBNZ, and my testing capabilities are limited on Android builds to find out.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_Branch.cpp:
```diff
@@ -74,6 +76,51 @@ void JitArm64::rfi(UGeckoInstruction inst)
+  SetJumpTarget(branch);
+}
+
+void JitArm64::WriteBranchWatchDestInScratch(u32 origin, Arm64Gen::ARM64Reg destination,
```
Nitpick: The "Scratch" in the function name here feels like an Jit64-ism. Perhaps replace "Scratch" with "Reg".

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_LoadStore.cpp:
```diff
@@ -829,6 +856,13 @@ void JitArm64::dcbx(UGeckoInstruction inst)
+    if (m_enable_debugging)
+    {
+      const auto& op = js.op[2];
+      WriteBranchWatch_dcbx(op.address, op.branchTo, op.inst, WA, gpr.GetCallerSavedUsed(),
+                            fpr.GetCallerSavedUsed());
```
Small optimization:

```suggestion
      WriteBranchWatch_dcbx(op.address, op.branchTo, op.inst, WA,
                            gpr.GetCallerSavedUsed() & ~BitSet32{WB}, fpr.GetCallerSavedUsed());
```

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_Branch.cpp:
```diff
@@ -74,6 +76,51 @@ void JitArm64::rfi(UGeckoInstruction inst)
+  ABI_CallFunction(
+      m_ppc_state.msr.IR ? &Core::BranchWatch::HitV_fk : &Core::BranchWatch::HitP_fk, branch_watch,
+      Common::BitCast<u64>(Core::FakeBranchWatchCollectionKey{origin, destination}), inst.hex);
+  ABI_PopRegisters(gprs_in_use);
+  m_float_emit.ABI_PushRegisters(fprs_in_use, WA);
```
```suggestion
  m_float_emit.ABI_PopRegisters(fprs_in_use, WA);
  ABI_PopRegisters(gprs_in_use);
```

JosJuice replied:
Registers must be popped in reverse order of how they were pushed. And you accidentally have a call to PushRegisters here instead of PopRegisters.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_LoadStore.cpp:
```diff
@@ -751,6 +753,31 @@ void JitArm64::stmw(UGeckoInstruction inst)
+      m_ppc_state.msr.IR ? &Core::BranchWatch::HitV_fk_n : &Core::BranchWatch::HitP_fk_n,
+      branch_watch, Common::BitCast<u64>(Core::FakeBranchWatchCollectionKey{origin, destination}),
+      inst.hex, n);
+  ABI_PopRegisters(gprs_in_use);
+  m_float_emit.ABI_PopRegisters(fprs_in_use, WB);
```
```suggestion
  m_float_emit.ABI_PopRegisters(fprs_in_use, WB);
  ABI_PopRegisters(gprs_in_use);
```

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_LoadStore.cpp:
```diff
@@ -751,6 +753,31 @@ void JitArm64::stmw(UGeckoInstruction inst)
+      inst.hex, n);
+  ABI_PopRegisters(gprs_in_use);
+  m_float_emit.ABI_PopRegisters(fprs_in_use, WB);
+  gpr.Unlock(WA);
+  gpr.Unlock(WB);
+  SetJumpTarget(branch);
```
```suggestion
  SetJumpTarget(branch);
  gpr.Unlock(WA);
  gpr.Unlock(WB);
```

In practice I think the emitted code will be okay even if you don't reorder these, but the code will be easier to reason about if you reorder them. Same for the other variants of this function. (Or well, I suppose this comment might not be relevant anymore once you fix what I said about not calling `Lock` inside conditionally executed code...)

mitaclaw commented on Source/Core/Core/PowerPC/CachedInterpreter/CachedInterpreter.cpp:
```diff
@@ -210,6 +211,15 @@ void CachedInterpreter::WriteBrokenBlockNPC(CachedInterpreter& cached_interprete
+{
+  if (auto& branch_watch = cached_interpreter.m_branch_watch; branch_watch.GetRecordingActive())
+  {
+    const PPCAnalyst::CodeOp& op = cached_interpreter.m_code_buffer[n.hex];
```
I have now realized that the Cached Interpreter JIT doesn't have OPTION_BRANCH_FOLLOW enabled, and since that is the only reason for this function to exist, it can be removed.

JosJuice replied:
That's true! It should be fine to change the `op.skip` case to just an assert that triggers on any instruction.

mitaclaw commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_Branch.cpp:
```diff
@@ -74,6 +76,51 @@ void JitArm64::rfi(UGeckoInstruction inst)
+{
+  // Use ARM64Reg::X0 as a temporary to avoid moving it later in ABI_CallFunction.
+  static constexpr ARM64Reg branch_watch = ARM64Reg::X0;
+  gpr.Lock(EncodeRegTo32(branch_watch));
```
Is my mistake that I put the `gpr.Unlock` call inside the scope of the conditional branch?

JosJuice replied:
No, the issue is deeper than that. `JitArm64::WriteBranchWatch` is called from code that contains conditional branches. Search for `FlushMode::MaintainState` in this file and you'll see the call sites.

dreamsyntax commented:
What is the status of this PR? Once conflicts are resolved, is this ready for a test run?

dreamsyntax commented:
@mitaclaw bumping, since this would be a shame to go stale for too long.

mitaclaw commented:
I apologize for my hiatus. I've resumed working on this tool in the last few days, and have added some new features:
![New Columns](https://github.com/dolphin-emu/dolphin/assets/140017135/c93023b0-ee30-446e-af18-a8a8954cbb08)
It is now possible to track code paths that execute only when a branch condition is false. Filter options for true-path and false-path branch hits are also available. I have also split the origin symbols and destination symbols into two columns. You may notice the dialog window has a lot going on. For this reason, I have also added support for hiding and moving columns:
![Hide Columns](https://github.com/dolphin-emu/dolphin/assets/140017135/39baa22a-fd60-4035-b31c-0ea200ce8464)
The hiding and moving of columns is saved to the Qt settings. JitArm64 fixes are still in-progress. I will push the changes as soon as they are ready.

JosJuice commented:
Yes, that could be worthwhile.

mitaclaw commented:
Do I need to update the copyright date of the newly committed files from 2023 to 2024?

lioncash commented:
They year only serves to establish a creation/origin date. So if they were created in 2023, it's sufficient to leave them as is.

mitaclaw commented:
I think almost everything has changed since the last force push. Here are some highlights:
- BranchWatchTableProxy.cpp/h in its entirety was moved into BranchWatchDialog.cpp, since nothing else should be using it anyway. I had trouble with the Q_OBJECT macro in a cpp file on Windows, so I had to remove it (there didn't seem to be any examples for a solution in the codebase).
BranchWatchTableView is now just a normal QTableView (this rewrite fulfilled a TODO in BranchWatchDialog.h, so that's good).
- I put in a last-minute feature to ignore branch hits during `CBoot::RunApploader`, but I wasn't really sure where to put the boolean flag. I am aware SConfig is in the process of being removed, so if there is a better place to put it, let me know.
- Added some tooltips in a few places in BranchWatchDialog to better explain things that may be vague for terseness.
- Revised the help text once again.
- JIT code should be *way* safer than it used to be. I gave up on the ABI_PARAM1 shortcut on x86_64 because I am no longer convinced it is safe at all on *any* platform.
I still need to do a search for obsoleted header includes and unrelated code edits.

mitaclaw commented:
I have cleaned up includes, reduced the LOC to just what is relevant, and solved a few lingering issues:
- Unscoped enums from BranchWatchTableModel.h that were formerly in the global namespace are now tucked away in a namespace, a neat pre-C++11 trick to keep enumerated constants from leaking out of scope. Qt loves using unscoped enums and implicitly converting to and from int, so I would prefer to keep them like this rather than peppering casts all over.
- Did some much-needed optimizing to allocate fewer registers in JitArm64 branch instructions, and cleaned up `JitArm64::dcbx` very slightly.
- Updated member and function names that reflected older iterations of the Branch Watch Tool's design.

I feel confident in this iteration of the PR, and am done modifying it unless requested.

mitaclaw commented:
I learned a way to run `QMenu::exec` at the mouse pointer without resorting to `QCursor::pos()`, so I guess I was lying about not touching the PR lol.

mitaclaw commented:
At @dreamsyntax's request, the multi-selection mode is now more functional, supporting instruction patching and address copying actions just like how it had worked when a single row was selected. This sort of fixed a possible design inconsistency as well (ctrl+right clicking an unselected row while only a single different row was selected would perform the action on the former rather than the latter), so I'm glad I implemented this change.

mitaclaw commented:
I had forgotten to update the help message to reflect the new context menu behavior. It's now fixed.

dreamsyntax reviewed (approved):
* Tested on Windows/Linux (x64) everything seems to be working well.

AdmiralCurtiss commented on Source/Core/Core/Debugger/BranchWatch.h:
```diff
@@ -0,0 +1,286 @@
+  return self = self | other;
+}
+
+class BranchWatchCollection final
+    : public std::unordered_map<BranchWatchCollectionKey, BranchWatchCollectionValue>
```
Subclassing STL types like this feels really icky... Is there an advantage to doing this over just `using BranchWatchCollection = std::unordered_map<BranchWatchCollectionKey, BranchWatchCollectionValue>`?

mitaclaw replied:
You can't forward declare a using declaration like you can a class / struct. I think in an earlier iteration of the design, I needed to forward declare some class from BranchWatch.h in another header to skip the include, and switched all of the using declarations over to subclassing to be consistent and future-proof. In my opinion, I prefer this way, but am not overly attached to it if it better fits the style to use using declarations.

AdmiralCurtiss commented on Source/Core/Core/Debugger/BranchWatch.h:
```diff
@@ -0,0 +1,286 @@
+  // All Hit member functions are for the CPUThread only
+  // HitXX_fk are optimized for when origin and destination can be passed in one register easily.
+  // HitXX_fk_n are the same, but also increment the total_hits by N (see dcbx JIT code).
+  static void HitVT_fk(BranchWatch* branch_watch, u64 fake_key, u32 inst)
```
All of these function names are not particularly self-explanatory. Can you maybe make them more clear or at least put some comments on what exactly they mean?

Also I assume they're static so that the JIT can call them easily? That should probably be in the comments here too.

mitaclaw replied:
Yes, these functions are static because I am unsure if `ABI_CallFunction` is compatible with the thiscall calling convention. I made the first parameter a pointer instead of a reference for a similar reason. Though I later found some JIT code which passes pointer parameters to functions expecting references, I'm not confident that's necessarily safe.
I think I disagree that the function names need further elaboration. I already explain the meaning behind the less obvious "fk" and "n" additions to the function name in the above comment and parameter names. The VT / VF / PT / PF portion of the function name can be figured out by looking at which collection member is being modified in the body of the function (self-documenting). These members have, in turn, comments at their declarations clarifying V / P is for Virtual / Physical, and T / F is for True Path / False Path. Must that be repeated everywhere VT / VF / PT / PF gets brought up?

mitaclaw replied:
Alright, I have caved and will offer two options to expand the abbreviated function names. The reason I shortened them in the first place was to keep things terse, and for that reason I prefer the commit named "Abbreviated" over the follow-up named "Full". Let me know which one is better.

AdmiralCurtiss replied:
Personally I prefer the Full over the Abbreviated, it's more clear at the call site without the direct context of seeing the BranchWatch internals next to it and the functions aren't used often enough to benefit from shortening IMO.

AdmiralCurtiss commented on Source/Core/Core/Debugger/BranchWatch.cpp:
```diff
@@ -0,0 +1,312 @@
+  m_blacklist_size = 0;
+}
+
+union USnapshotMetadata
```
This could also use a comment for what exactly it's designed for.

mitaclaw replied:
I have now added one.

AdmiralCurtiss replied:
Yeah that's a lot better, thanks.

AdmiralCurtiss commented on Source/Core/Core/Debugger/BranchWatch.cpp:
```diff
@@ -0,0 +1,312 @@
+      m_selection.emplace_back(BranchWatchSelectionValueType{&*kv_iter, is_virtual, condition,
+                                                             snapshot_metadata.inspection});
+    }
+    else if (hits_snapshot != 0)
+      ++m_blacklist_size;  // This will be very wrong when not in Blacklist mode. That's ok.
```
This needs enclosing braces per our coding guidelines.

AdmiralCurtiss commented on Source/Core/Core/Debugger/BranchWatch.h:
```diff
@@ -0,0 +1,286 @@
+  }
+
+private:
+  static Collection BranchWatch::*FindCollectionV(bool condition)
+  {
+    if (condition)
+      return &BranchWatch::m_collection_vt;
+    return &BranchWatch::m_collection_vf;
+  }
+
+  static Collection BranchWatch::*FindCollectionP(bool condition)
+  {
+    if (condition)
+      return &BranchWatch::m_collection_pt;
+    return &BranchWatch::m_collection_pf;
+  }
+
+  static Collection BranchWatch::*FindCollection(bool is_virtual, bool condition)
+  {
+    if (is_virtual)
+      return FindCollectionV(condition);
+    return FindCollectionP(condition);
+  }
```
Is there a reason these are static? It looks to me as if these are used in exactly one place, in a member function that already has the this pointer. So these could just return a `Collection&` and be much more standard syntax-wise.

AdmiralCurtiss commented on Source/Core/Core/Debugger/BranchWatch.cpp:
```diff
@@ -0,0 +1,312 @@
+  if (file == nullptr)
+    return;
+
+  const auto save_collection = [&](Collection BranchWatch::*collection, bool is_virtual,
+                                   bool condition) {
```
Likewise here, why can you not just pass a `const Collection&` and get rid of the somewhat confusing pointer-to-member? There's a few more instances of this pattern in this file, too.

mitaclaw replied:
Sometimes I get a little ahead of myself finding uses for lesser-used sects of the C++ language. You are right, I can just use Collection references in a few places.

AdmiralCurtiss reviewed (commented):
This seems largely fine but you have a few things that I would at least like a comment for.

It would be good if @JosJuice could sign off on the JIT code changes, too.

AdmiralCurtiss reviewed (approved):
Untested, but this seems reasonable enough to me. I have not reviewed the JIT changes, so please do before merging!

mitaclaw commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_Branch.cpp:
```diff
@@ -108,22 +183,41 @@ void JitArm64::bx(UGeckoInstruction inst)
+  {
+    BitSet32 gpr_caller_save = gpr.GetCallerSavedUsed();
+    if (WA != ARM64Reg::INVALID_REG)
+      gpr_caller_save &= ~BitSet32{DecodeReg(WA)};
```
I feel like I've made a mistake here, masking out WA from the gpr_caller_save.

mitaclaw commented:
For any remaining changes, I will be waiting to merge them with the existing five commits until the review process is finished to avoid confusion about whether things already approved need re-checked.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_LoadStore.cpp:
```diff
@@ -769,29 +770,26 @@ void JitArm64::dcbx(UGeckoInstruction inst)
+    BitSet32 gpr_caller_save;
+    if (IsDebuggingEnabled())
+      gpr_caller_save = gpr.GetCallerSavedUsed() & ~BitSet32{DecodeReg(WB)};
+    const ARM64Reg reg_cycle_count = gpr.GetReg(), reg_downcount = gpr.GetReg();
```
GetCallerSavedUsed needs to come after GetReg.

mitaclaw commented on Source/Core/Core/PowerPC/Jit64/Jit_Branch.cpp:
```diff
@@ -204,6 +290,12 @@ void Jit64::bcctrx(UGeckoInstruction inst)
+    if (IsDebuggingEnabled())
+    {
+      const BitSet32 caller_save = CallerSavedRegistersInUse() | BitSet32{RSCRATCH};
+      const RCX64Reg bw_reg_b = gpr.Scratch();
```
I suspected what you have said about `RegCache::Scratch()` recycling a register that is caller saved, but figured it was an unlikely missed optimization and not inherently unsafe. However, your advice for what is safe after a GPR flush is something I had not figured out, so thank you.

JosJuice replied:
Technically it isn't unsafe – you're just pushing a register that you don't necessarily need to push – but it is a bug, and we want the JIT to be as performant as possible.

RSCRATCH_EXTRA is a register you can normally only use if you asked the register cache to reserve it for you, but since you've flushed all registers, you can just go ahead and use it without asking the register cache.

mitaclaw commented on Source/Core/Core/PowerPC/Jit64/Jit_LoadStore.cpp:
```diff
@@ -258,9 +259,12 @@ void Jit64::dcbx(UGeckoInstruction inst)
+    loop_counter = gpr.Scratch();
+    BitSet32 bw_caller_save;
+    if (IsDebuggingEnabled())
+      bw_caller_save = CallerSavedRegistersInUse() & BitSet32{RSCRATCH2};
```
Ack, how did I let that `&` slip through?

mitaclaw commented on Source/Core/Core/ConfigManager.h:
```diff
@@ -45,6 +45,7 @@ struct SConfig
 {
   // Settings
   bool bBootToPause = false;
+  bool bBranchWatchIgnoreApploader = false;
```
Thanks. As I said, I just wasn't sure where the best place to put this was. I might make a PR to move bBootToPause into System as well after this is done.

mitaclaw commented on Source/Core/Core/Debugger/BranchWatch.h:
```diff
@@ -0,0 +1,277 @@
+        .total_hits += n;
+  }
+
+  static void HitPhysicalTrue_fk_n(BranchWatch* branch_watch, u64 fake_key, u32 inst, u32 n)
```
I have updated the comment to make the meaning of "fk" less subtle.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_Branch.cpp:
```diff
@@ -108,22 +183,35 @@ void JitArm64::bx(UGeckoInstruction inst)
+  if (IsDebuggingEnabled())
+  {
+    const ARM64Reg WB = gpr.GetReg(), WC = gpr.GetReg();
+    WriteBranchWatch<true>(js.compilerPC, js.op->branchTo, inst, WB, WC, {}, {});
```
You need to save WA here if inst.LK is true.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_Branch.cpp:
```diff
@@ -231,6 +348,12 @@ void JitArm64::bcctrx(UGeckoInstruction inst)
+  if (IsDebuggingEnabled())
+  {
+    const ARM64Reg WC = gpr.GetReg(), WD = gpr.GetReg();
+    WriteBranchWatchDestInRegister(js.compilerPC, WA, inst, WC, WD, {}, {});
```
You need to save WA here, and also WB if inst.LK_3 is true.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_Branch.cpp:
```diff
@@ -281,6 +406,18 @@ void JitArm64::bclrx(UGeckoInstruction inst)
 
+  if (IsDebuggingEnabled())
+  {
+    BitSet32 gpr_caller_save = {};
+    BitSet32 fpr_caller_save = {};
+    if (conditional)
+    {
+      gpr_caller_save = gpr.GetCallerSavedUsed() & ~BitSet32{DecodeReg(WB), DecodeReg(WC)};
+      fpr_caller_save = fpr.GetCallerSavedUsed();
+    }
```
You need to save WA unless js.op->branchIsIdleLoop is true.

I suggest doing it as follows:

```suggestion
    BitSet32 gpr_caller_save = gpr.GetCallerSavedUsed() & ~BitSet32{DecodeReg(WB), DecodeReg(WC)};
    BitSet32 fpr_caller_save = fpr.GetCallerSavedUsed();
    if (js.op->branchIsIdleLoop)
      gpr_caller_save[DecodeReg(WA)] = false;
```

Note that flushing with FlushMode::All flushes all guest registers, not all host registers. So the register cache will still consider WA to be in use until you call Unlock(WA).

mitaclaw replied:
Isn't it still worth doing less work if conditional is true? Let me know if my latest commits are still missing the mark here.

JosJuice replied:
Ah, you mean skipping the work of calling `GetCallerSavedUsed`? That's true, that might be worth skipping. I was thinking more about the source code looking messy. Feel free to do it either way, but if you skip calling GetCallerSavedUsed in the `!conditional` case, make sure to manually set `WA` to saved when `!js.op->branchIsIdleLoop`.
Perhaps, instead of my suggestion of `if (js.op->branchIsIdleLoop) gpr_caller_save[DecodeReg(WA)] = false;`, you could do `gpr_caller_save[DecodeReg(WA)] = !js.op->branchIsIdleLoop;`.

JosJuice replied:
Ah, I see now you already implemented it. Your current implementation seems fine.

JosJuice commented on Source/Core/Core/PowerPC/Jit64/Jit_Branch.cpp:
```diff
@@ -160,7 +164,8 @@ void Jit64::bx(UGeckoInstruction inst)
   if (IsDebuggingEnabled())
   {
-    WriteBranchWatch<true>(js.compilerPC, js.op->branchTo, inst, RSCRATCH, RSCRATCH2, {});
+    static_assert((BitSet32{ABI_PARAM1} & ABI_ALL_CALLER_SAVED) != BitSet32{});
```
What is the purpose of this static assert? Maybe I'm missing something, but I don't see why it would matter if ABI_PARAM1 is caller-saved.

mitaclaw replied:
I have probably just misunderstood how to document that this optimization is only safe after a GPR flush. I can replace these with comments saying this more directly?

JosJuice replied:
Okay, I understand then. The static assert didn't actually check whether the GPRs had been flushed – that would have required a runtime assert. The new comment looks fine.

JosJuice commented:
Interpreter and JIT changes LGTM. Please clean up the commit history before merge.

mitaclaw commented:
Commits squashed, rebased on latest master.

dreamsyntax reviewed (approved):
* Tested on a M1 Mac (ARM64/Apple Silicon) (JIT)
* Tested on X64 Linux & Windows (JIT, Cached Interp, Interp)

Everything feels good to me.

mitaclaw commented:
I just had the realization that the Branch Watch Dialog would remain up when disabling the debugging UI. Thankfully, it doesn't seem like anything breaks from it interacting with the hidden Code View Widget, nor from patching instructions in memory. Predictably, it fails to see any recompiled branch hits since that code is not being emitted (this is a solid way to confirm that). Still, the dialog remaining up, will that be a problem?

JosJuice commented:
I wouldn't consider it a problem.

mitaclaw commented:
I realized symbol filtering in Branch Watch Dialog was case-sensitive, unlike Code Widget. The fix was a minor change to `BranchWatchProxyModel::filterAcceptsRow`.

JosJuice commented:
@AdmiralCurtiss OK to merge?
